### PR TITLE
feat: Implement `privateFieldsAsSymbols` assumption for classes

### LINKED ISSUE #15389: [Bug]: Assumption privateFieldsAsProperties w/o transform-runtime plugin can lead to private field name clashes
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

When using the assumption `privateFieldsAsProperties: true` in combination with not using the Babel `transform-runtime` plugin (so that babel-helpers are inlined), it becomes possible to override #-private fields, which is incorrect semantics.

The issue can only be reproduced when using at least two different source files.

Bar.js:
```js
export default class Bar {
  #hidden = "Bar's hidden";
  getHidden() {
    return this.#hidden;
  }
}
```

index.js:
```js
import Bar from "./Bar";

class Foo extends Bar {
  // it should not be possible to override a #private field of a superclass:
  #hidden = "Foo's hidden";
}

console.log(new Foo().getHidden()); // actual: "Foo's hidden", expected: "Bar's hidden"
```

Since Babel's own REPL only supports one input file, here is the full example including a `.babelrc` in  codesandbox:
[codesandbox.io reproducer](https://codesandbox.io/s/babel-privatefieldsasproperties-bug-gqvfsz?file=/.babelrc:84-88)

### Configuration file name

.babelrc

### Configuration

```json
{
  "presets": [
    "env"
  ],
  "assumptions": {
    "privateFieldsAsProperties": true
  }
}
```

### Current and expected behavior

When running the codesandbox example given above, in the output you see `"Foo's hidden"`, so class `Foo` managed to override the #-private field `#hidden` of class `Bar`, which must not be possible.
When you switch off `privateFieldsAsProperties` and run the example again, the correct result `"Foo's hidden"` is shown.
Also when you use the Babel plugin `transform-runtime`, the behavior is correct.

### Environment

See [codesandbox example](https://codesandbox.io/s/babel-privatefieldsasproperties-bug-gqvfsz?file=/package.json):
- Babel 7.2.0
- Node 16

### Possible solution

The reason that both #-private fields share the same name `__private_0_hidden` is that the helper function [`classPrivateFieldLooseKey`](https://github.com/babel/babel/blob/main/packages/babel-helpers/src/helpers.ts#L997-L1002) tries to define and initialize a global counter variable `id` that is increased for each generated property name, to ensure unique private property names. However, when helper functions are inlined (no `transform-runtime` plugin), each bundled file comes with a _copy_ of this helper code, each defining its own `var id = 0;`. Thus, in both bundled files, the private member counter starts with `0`, ending up in the same private field slot name in both the superclass and the subclass, which must not happen.

One approach to solve this could be to use a `Symbol` instead of a generated string slot name, but I don't know if Babel helper code is allowed to use language features that may not be available in older target environments. I tested this approach and it works perfectly (given that `Symbol` is supported).
I guess helper code is not compiled, so [using `Symbol` would rule out IE](https://caniuse.com/?search=Symbol), unless a [`Symbol` polyfill](https://github.com/medikoo/es6-symbol) is applied. If this is a show-stopper, a new assumption `privateFieldsAsSymbolProperties` could be introduced that fixes the issue, but requires a target that supports `Symbol` (at least through a polyfill).

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by fwienber**
Issue #15389: When babel-helpers are inlined, `classPrivateFieldLooseKey()` defines a global variable `id` and initialized it to zero in every generated file. This can lead to name clashes between #-private members in subclass and superclass, which again leads to incorrect semantics.

It has been discussed to fix this (see my previous PR #15393), but eventually the suggestion was to introduce a new assumption `privateFieldsAsSymbols` and deprecate (and later remove) the buggy `privateFieldsAsProperties`.
In the new solution, instead of trying to produce a globally unique private property name string, a `Symbol` is created. To use the new assumption, in environments where `Symbol` is not available (Internet Explorer), a polyfill must be included.

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #15389
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2721
| Any Dependency Changes?  | No
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15435"></a>

### CODE REVIEW

**Comment by fwienber** [comment]
Okay, I seem to have used a wrong type import, which lead to weird CI failures.
But now all tests are green.
Currently, I just duplicated _one_ test that only tests a #-private instance method.
I was wondering whether I have to duplicate _all_ tests that now use `privateFieldsAsProperties` to also run with `privateFieldsAsSymbols`. If not, maybe a few more that cover all cases (private instance field/method, static field/method, no name-clash when inheriting) would be nice. Or would we change the existing tests on the Babel 8 branch (if there even is such a thing)?

File: packages/babel-helper-create-class-features-plugin/src/index.ts [hunk 1]
```diff
@@ -61,6 +64,9 @@ export function createClassFeaturePlugin({
     if (privateFieldsAsProperties !== undefined) {
       explicit.push(`"privateFieldsAsProperties"`);
     }
+    if (privateFieldsAsSymbols !== undefined) {
+      explicit.push(`"privateFieldsAsSymbols"`);
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/index.ts:68]: Nit: we can export the `AssumptionName` type and annotate `explicit` as <code>&grave;"${AssumptionName}"&grave;[]</code>.

**Review by nicolo-ribaudo — APPROVED** [review]
Thank you!

I pushed some changes instead of just leaving comments, because we are preparing the next minor release and I'd like to have this PR ready as soon as possible.

- I improved the type annotation as @JLHwung suggested
- I added an error if both `privateFieldsAsSymbols` and `privateFieldsAsProperties` are enabled at the same time
- I duplicated all the existing `privateFieldsAsProperties` tests

File: packages/babel-helper-create-class-features-plugin/src/index.ts (hunk 1 above)
Comment (fwienber) [reply packages/babel-helper-create-class-features-plugin/src/index.ts:76]: Ah, so the (re-)export of `AssumptionName` was missing! And adding the quotes to the type is a nice trick! In my first attempt, I just added them afterwards (during output).

File: packages/babel-helper-create-class-features-plugin/src/fields.ts [hunk 2]
```diff
@@ -80,15 +84,18 @@ export function buildPrivateNamesNodes(
...
       ]);
+      annotateAsPure(init);
+    } else if (privateFieldsAsSymbols) {
+      init = t.callExpression(t.identifier("Symbol"), [t.stringLiteral(name)]);
     } else if (!isStatic) {
       init = t.newExpression(
         t.identifier(!isMethod || isAccessor ? "WeakMap" : "WeakSet"),
         [],
       );
+      annotateAsPure(init);
```
Comment (fwienber) [inline packages/babel-helper-create-class-features-plugin/src/fields.ts:95]: This is like before, but while at it, does `new WeakMap()` / `new WeakSet()` really need a PURE annotation?

Comment (nicolo-ribaudo) [reply packages/babel-helper-create-class-features-plugin/src/fields.ts:95]: Oh actually, it looks like the `PURE` annotations is also needed for symbols:
Terser playground with minifying `let x = /*#__PURE__*/ Symbol("x");let y = Symbol("y");`

Comment (fwienber) [reply packages/babel-helper-create-class-features-plugin/src/fields.ts]: Since this has the same effect as `new Symbol("y")`, I can see why.

**Pull request merged**